### PR TITLE
pre-check if uri is parsable or we're getting something bad from the cli...

### PR DESCRIPTION
**Pull request by eishay**
...ent before passing it into the commander layer
@andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/ext/ExtPageController.scala [hunk 1]
```diff
@@ -31,8 +32,14 @@ class ExtPageController @Inject() (
...
     if (url.isEmpty) throw new Exception(s"empty url for json ${request.body} for user ${request.user}")
-    val info = pageCommander.getPageDetails(url, request.userId, request.experiments)
-    Ok(Json.toJson(info))
+    URI.parse(url) match {
+      case Success(_) =>
+        val info = pageCommander.getPageDetails(url, request.userId, request.experiments)
+        Ok(Json.toJson(info))
+      case Failure(e) =>
+        log.error(s"Error parsing url: $url", e)
+        BadRequest(s"Error parsing url: $url")
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/ext/ExtPageController.scala]: Mind making this JSON?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/mobile/MobilePageController.scala [hunk 2]
```diff
@@ -22,8 +25,14 @@ class MobilePageController @Inject() (
...
     val url = (request.body \ "url").as[String]
-    val info = pageCommander.getPageDetails(url, request.userId, request.experiments)
-    Ok(Json.toJson(info))
+    URI.parse(url) match {
+      case Success(_) =>
+        val info = pageCommander.getPageDetails(url, request.userId, request.experiments)
+        Ok(Json.toJson(info))
+      case Failure(e) =>
+        log.error(s"Error parsing url: $url", e)
+        BadRequest(s"Error parsing url: $url")
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/mobile/MobilePageController.scala]: Same, I assume mobile would blow up on non-json

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/ext/ExtPageController.scala (hunk 1 above)
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/ext/ExtPageController.scala]: sure, thought i would expect the client to first check the response status before parsing and not try to parse if its not a 200.

**Comment by eishay** [comment]
@andrewconner like this?

**Comment by andrewconner** [comment]
Perfect thanks

**Comment by 2is10** [comment]
I think it’s sometimes useful to get reports like this, so that we’ll know how we can improve our parser. In this case, for example, Chrome said the fragment part contains a `#`. Chrome probably permits an unescaped `#` in the fragment because it doesn’t cause any ambiguity about the structure of the URL. The fragment begins at the first `#`. I think our parser would be better if it tolerated a `#` in a fragment, especially in light of the fact that Chrome does.

**Comment by eishay** [comment]
- @ymatsuda 
  The reason why I added the intolerance for `#` was because we eventually use the apache http client which parse the url using the java URL class, and that one blows up.
  Another option is to let the URL in and know that we can't fetch content using it with the current libraries we use.

**Comment by 2is10** [comment]
Two observations: 1) The fragment is not sent in an HTTP request, so it can be discarded before constructing a `java.net.URL`. 2) If we do want to correct it for whatever reason, escaping `#` to `%23` would produce an equivalent, conforming URL.

**Comment by eishay** [comment]
Didn't know about #1, I'll use it. We'll also do the escaping instead of just throwing exception. More pull requests will follow.

**Pull request merged**